lib/iris/tests/test_concatenate.py

```diff
@@ -352,52 +352,72 @@ class Test2D(tests.IrisTest):
         y = (0, 2)
-        cube = _make_cube((0, 2), y, 1, mask=True)
+        fill_value = 1234
+        cube = _make_cube((0, 2), y, 1, mask=True,
+                          fill_value=fill_value)
         cubes.append(cube)
         cubes.append(_make_cube((2, 4), y, 2))
         result = concatenate(cubes)
         self.assertEqual(len(result), 1)
+        self.assertEqual(result[0].fill_value, fill_value)
+        self.assertEqual(result[0].data.fill_value, fill_value)
 
     def test_masked_and_unmasked_int16(self):
         cubes = []
         y = (0, 2)
         dtype = np.dtype('int16')
-        cube = _make_cube((0, 2), y, 1, dtype=dtype, mask=True)
+        fill_value = 1234
+        cube = _make_cube((0, 2), y, 1, dtype=dtype, mask=True,
+                          fill_value=fill_value)
         cubes.append(cube)
         cubes.append(_make_cube((2, 4), y, 2, dtype=dtype))
         result = concatenate(cubes)
         self.assertEqual(len(result), 1)
+        self.assertEqual(result[0].fill_value, fill_value)
+        self.assertEqual(result[0].data.fill_value, fill_value)
 
     def test_unmasked_and_masked(self):
         cubes = []
         y = (0, 2)
         cubes.append(_make_cube((0, 2), y, 1))
-        cube = _make_cube((2, 4), y, 2, mask=True)
+        fill_value = 1234
+        cube = _make_cube((2, 4), y, 2, mask=True,
+                          fill_value=fill_value)
         cubes.append(cube)
         result = concatenate(cubes)
         self.assertEqual(len(result), 1)
+        self.assertEqual(result[0].fill_value, fill_value)
+        self.assertEqual(result[0].data.fill_value, fill_value)
 
     def test_unmasked_and_masked_int16(self):
         cubes = []
         y = (0, 2)
         dtype = np.dtype('int16')
         cubes.append(_make_cube((0, 2), y, 1, dtype=dtype))
-        cube = _make_cube((2, 4), y, 2, dtype=dtype, mask=True)
+        fill_value = 1234
+        cube = _make_cube((2, 4), y, 2, dtype=dtype, mask=True,
+                          fill_value=fill_value)
         cubes.append(cube)
         result = concatenate(cubes)
         self.assertEqual(len(result), 1)
+        self.assertEqual(result[0].fill_value, fill_value)
+        self.assertEqual(result[0].data.fill_value, fill_value)
 
     def test_masked_with_data_fill_value_difference(self):
         cubes = []
         y = (0, 2)
-        cube = _make_cube((0, 2), y, 1, mask=True)
-        cube.data.fill_value = 10
+        fill_value = 1234
+        cube = _make_cube((0, 2), y, 1, mask=True,
+                          fill_value=fill_value)
+        cube._data_manager._real_array.fill_value = 10
         cubes.append(cube)
         cube = _make_cube((2, 4), y, 1, mask=True)
-        cube.data.fill_value = 20
+        cube._data_manager._real_array.fill_value = 20
         cubes.append(cube)
         result = concatenate(cubes)
         self.assertEqual(len(result), 1)
+        self.assertEqual(result[0].fill_value, fill_value)
```

Comment:
@dkillick Look again :eyes:  

The concatenation should fail if the `fill_value` on each of the cube are both non-`None` and different. That's **not** the case. The `fill_values` are **different** on the actual masked arrays. We do not care about that difference!

Infact, for this test case (subtle as it is) highlights that the actual masked array `fill_value` does not matter. Here once cube does have its `fill_value` metadata set, the other not, so they concatenate based on `fill_value` promotion!

This test is good.